File: car.lua
-- Car factory and physics (pure logic, no Love2D dependency)

local damage = require("damage")

local Car = {}
Car.__index = Car
local abs, min, max, cos, sin = math.abs, math.min, math.max, math.cos, math.sin
local random, pi = math.random, math.pi

function Car.new(track, overrides)
    local self = setmetatable({}, Car)
    overrides = overrides or {}

    -- Start position (with optional offset along track)
    self.x = track.startX or track.cx
    self.y = track.startY or (track.cy - 50)
    self.angle = track.startAngle or 0
    self.speed = 0
    self.width = 28
    self.height = 14
    self.prevSpeed = 0
    self.turning = false
    self.currentZone = nil
    self.shouldSpawnSmoke = false
    self.shouldSpawnDarkSmoke = false  -- engine damage smoke

    -- Identity
    self.name = overrides.name or "Player"
    self.color = overrides.color or {0.85, 0.1, 0.1}
    self.isAI = overrides.isAI or false

    self.physics = {
        mass = 800,
        fuelMass = 50,
        maxFuel = 50,
        fuelRate = 1.5,
        tirePressure = 2.2,
        optimalPressure = 2.2,
        engineForce = 250000,
        brakeForce = 200000,
        dragCoeff = 3.0,
        rollingResistance = 0.015,
        maxSpeed = 320,
        baseTurnSpeed = 3.0,
        gripMultiplier = 1.0,
        bumpMultiplier = 1.0,
    }

    -- Apply physics overrides
    if overrides.physics then
        for k, v in pairs(overrides.physics) do
            self.physics[k] = v
        end
    end

    -- Apply start offset along track if provided
    if overrides.startOffset and track.getPointAtPercent then
        local offsetPct = (1.0 + overrides.startOffset) % 1.0
        local pt = track.getPointAtPercent(offsetPct)
        if pt then
            self.x = pt.x
            self.y = pt.y
        end
    end

    -- Apply lateral offset (perpendicular to track direction) for grid starts
    if overrides.lateralOffset and overrides.lateralOffset ~= 0 then
        local perpAngle = self.angle + pi / 2
        self.x = self.x + cos(perpAngle) * overrides.lateralOffset
        self.y = self.y + sin(perpAngle) * overrides.lateralOffset
    end

    -- Damage state (always present)
    self.damage = damage.create()

    return self
end

function Car:update(dt, input, track)
    local physics   = self.physics
    local totalMass = physics.mass + physics.fuelMass

    local zone    = track.getSurfaceAt(self.x, self.y)
    self.currentZone = zone
    local onTrack = track.isOnTrack(self.x, self.y)

    -- ----------------------------------------------------------------
    -- Damage modifiers
    -- ----------------------------------------------------------------
    local dmgMods = damage.getHandlingModifiers(self.damage)

    -- Tire pressure grip
    local pressureDev  = abs(physics.tirePressure - physics.optimalPressure)
    local pressureGrip = max(0.3, 1.0 - pressureDev * 0.4)

    -- Effective grip (also reduced by average tire health)
    local surfaceGrip   = onTrack and zone.grip or 0.3
    local effectiveGrip = surfaceGrip * pressureGrip * physics.gripMultiplier
                        * dmgMods.avgTireHealth
    effectiveGrip = min(1.0, max(0.1, effectiveGrip))

    -- Bumpiness: suspension damage amplifies it; flat tires add periodic thump
    local baseBump = onTrack and (zone.bumpiness * physics.bumpMultiplier) or 0.0
    local bumpiness = baseBump * dmgMods.bumpMult

    -- Flat-tire periodic thumping: each flat wheel adds low-frequency perturbation
    local flatCount = damage.flatTireCount(self.damage)
    if flatCount > 0 and abs(self.speed) > 15 then
        -- Thump frequency scales with speed (like wheel hitting rim each revolution)
        local thumpFreq = abs(self.speed) / 60   -- ~1 thump/sec at speed 60
        local thumpPhase = (love and love.timer and love.timer.getTime() or 0) * thumpFreq
        local thump = max(0, sin(thumpPhase * 2 * pi))
        bumpiness = bumpiness + thump * flatCount * 0.35
    end

    self.prevSpeed = self.speed

    -- Throttle / brake
    local throttle = 0
    if input.up and physics.fuelMass > 0 then
        throttle = 1
    end
    local braking = input.down

    -- Engine damage reduces drive force
    local driveForce = throttle * physics.engineForce * effectiveGrip * dmgMods.engineMult
    local brakeDecel = 0
    if braking then
        brakeDecel = physics.brakeForce * effectiveGrip
    end

    -- Drag (body damage increases drag)
    local dragForce = physics.dragCoeff * self.speed * abs(self.speed) * dmgMods.dragMult

    -- Rolling resistance
    local rollingForce = physics.rollingResistance * totalMass * 9.81

    -- Off-track grass drag
    local grassDrag = 0
    if not onTrack then
        grassDrag = abs(self.speed) * 3.0
    end

    -- Net force
    local netForce = driveForce - dragForce - rollingForce - grassDrag
    if braking then
        if self.speed > 0 then
            netForce = netForce - brakeDecel
        elseif self.speed < 0 then
            netForce = netForce + brakeDecel
        else
            netForce = netForce - brakeDecel * 0.3
        end
    end

    local accel = netForce / totalMass
    self.speed  = self.speed + accel * dt

    -- Bumpiness perturbation (suspension damage amplified)
    if bumpiness > 0.01 and abs(self.speed) > 20 then
        local bumpMag = bumpiness * abs(self.speed) * 0.0003
        self.speed = self.speed + (random() - 0.5) * bumpMag * self.speed
        self.angle = self.angle + (random() - 0.5) * bumpiness * 0.005
    end

    -- Clamp speed (max speed reduced by damage)
    local effectiveMaxSpeed = physics.maxSpeed * dmgMods.maxSpeedMult
    self.speed = max(-100, min(effectiveMaxSpeed, self.speed))

    -- Stop drifting at low speeds
    if abs(self.speed) < 1 and throttle == 0 and not braking then
        self.speed = 0
    end

    -- Fuel consumption
    if throttle > 0 then
        physics.fuelMass = max(0, physics.fuelMass - physics.fuelRate * dt)
    end

    -- Turning (supports both boolean left/right and continuous steer)
    local turnFactor = min(1, abs(self.speed) / 100) * effectiveGrip
    self.turning = false
    if input.steer then
        local steerVal = max(-1, min(1, input.steer))
        if abs(steerVal) > 0.05 then
            self.angle   = self.angle + physics.baseTurnSpeed * steerVal * turnFactor * dt
            self.turning = true
        end
    else
        if input.left then
            self.angle   = self.angle - physics.baseTurnSpeed * turnFactor * dt
            self.turning = true
        end
        if input.right then
            self.angle   = self.angle + physics.baseTurnSpeed * turnFactor * dt
            self.turning = true
        end
    end

    -- Tire pull: asymmetric damage pulls the car to one side
    -- Player must compensate with steering (NPCs experience it through physics)
    if abs(dmgMods.tirePull) > 0.01 and abs(self.speed) > 10 then
        local pullTurn = dmgMods.tirePull * physics.baseTurnSpeed * 0.35 * turnFactor
        self.angle = self.angle + pullTurn * dt
    end

    -- Move car
    self.x = self.x + cos(self.angle) * self.speed * dt
    self.y = self.y + sin(self.angle) * self.speed * dt

    -- Keep in bounds
    self.x = max(10, min(790, self.x))
    self.y = max(10, min(590, self.y))

    -- Smoke flags
    local isBraking    = input.down and self.speed > 50
    local isSharpTurn  = self.turning and abs(self.speed) > 120
    self.shouldSpawnSmoke     = isBraking or isSharpTurn
    -- Dark engine-damage smoke when engine is hurt and car is moving
    self.shouldSpawnDarkSmoke = self.damage and self.damage.engine < 0.55
                                and abs(self.speed) > 25
end

return Car

File: crates/sim/src/car.rs
//! Car factory and physics (port of car.lua).

use crate::damage::{self, DamageState};
use crate::nnet::{Net, NetData};
use crate::npc_profiles::Personality;
use crate::rng::GameRng;
use crate::track::Track;

#[derive(Debug, Clone, Copy)]
pub struct Physics {
    pub mass: f64,
    pub fuel_mass: f64,
    pub max_fuel: f64,
    pub fuel_rate: f64,
    pub tire_pressure: f64,
    pub optimal_pressure: f64,
    pub engine_force: f64,
    pub brake_force: f64,
    pub drag_coeff: f64,
    pub rolling_resistance: f64,
    pub max_speed: f64,
    pub base_turn_speed: f64,
    pub grip_multiplier: f64,
    pub bump_multiplier: f64,
}

impl Default for Physics {
    fn default() -> Self {
        Self {
            mass: 800.0,
            fuel_mass: 50.0,
            max_fuel: 50.0,
            fuel_rate: 1.5,
            tire_pressure: 2.2,
            optimal_pressure: 2.2,
            engine_force: 250_000.0,
            brake_force: 200_000.0,
            drag_coeff: 3.0,
            rolling_resistance: 0.015,
            max_speed: 320.0,
            base_turn_speed: 3.0,
            grip_multiplier: 1.0,
            bump_multiplier: 1.0,
        }
    }
}

/// Per-profile physics overrides (only the fields a profile customizes).
#[derive(Debug, Clone, Copy, Default)]
pub struct PhysicsOverrides {
    pub mass: Option<f64>,
    pub engine_force: Option<f64>,
    pub brake_force: Option<f64>,
    pub base_turn_speed: Option<f64>,
    pub grip_multiplier: Option<f64>,
    pub max_speed: Option<f64>,
}

impl PhysicsOverrides {
    pub fn apply_to(&self, p: &mut Physics) {
        if let Some(v) = self.mass {
            p.mass = v;
        }
        if let Some(v) = self.engine_force {
            p.engine_force = v;
        }
        if let Some(v) = self.brake_force {
            p.brake_force = v;
        }
        if let Some(v) = self.base_turn_speed {
            p.base_turn_speed = v;
        }
        if let Some(v) = self.grip_multiplier {
            p.grip_multiplier = v;
        }
        if let Some(v) = self.max_speed {
            p.max_speed = v;
        }
    }
}

/// Per-frame control input. `steer` (continuous, [-1, 1]) takes precedence
/// over the boolean left/right when set — same contract as the Lua code.
#[derive(Debug, Clone, Copy, Default)]
pub struct CarInput {
    pub up: bool,
    pub down: bool,
    pub left: bool,
    pub right: bool,
    pub steer: Option<f64>,
}

/// Evolution + driving-error state carried by AI cars.
#[derive(Debug, Clone)]
pub struct NpcState {
    pub brain: Net,
    pub personality: Personality,
    pub best_brain: NetData,
    pub best_fitness: f64,
    pub generation: u32,
    pub current_fitness: f64,

    // Per-race metrics (see ai::init_metrics)
    pub time_off_track: f64,
    pub time_stationary: f64,
    pub avg_speed: f64,
    pub speed_samples: f64,
    pub stuck_timer: f64,
    pub stuck_override: f64,
    pub stuck_steer_dir: f64,

    // Error state for imperfect driving
    pub lapse_timer: f64,
    pub last_input: Option<CarInput>,
}

#[derive(Debug, Clone, Default)]
pub struct CarOverrides {
    pub name: Option<String>,
    pub color: Option<[f32; 3]>,
    pub is_ai: bool,
    pub physics: PhysicsOverrides,
    /// Offset along the track (fraction of a lap, may be negative) for grid starts.
    pub start_offset: Option<f64>,
    /// Lateral offset perpendicular to the track direction.
    pub lateral_offset: f64,
}

#[derive(Debug, Clone)]
pub struct Car {
    pub x: f64,
    pub y: f64,
    pub angle: f64,
    pub speed: f64,
    pub width: f64,
    pub height: f64,
    pub prev_speed: f64,
    pub turning: bool,
    pub current_zone_name: String,
    pub should_spawn_smoke: bool,
    pub should_spawn_dark_smoke: bool,

    pub name: String,
    pub color: [f32; 3],
    pub is_ai: bool,

    pub physics: Physics,
    pub damage: DamageState,
    pub npc: Option<NpcState>,
}

impl Car {
    pub fn new(track: &Track, overrides: CarOverrides) -> Self {
        let mut physics = Physics::default();
        overrides.physics.apply_to(&mut physics);

        let mut x = track.start_x;
        let mut y = track.start_y;
        let angle = track.start_angle;

        // Start offset along track
        if let Some(offset) = overrides.start_offset {
            let offset_pct = (1.0 + offset).rem_euclid(1.0);
            let pt = track.get_point_at_percent(offset_pct);
            x = pt.x;
            y = pt.y;
        }

        // Lateral offset perpendicular to track direction (for grid starts)
        if overrides.lateral_offset != 0.0 {
            let perp = angle + std::f64::consts::FRAC_PI_2;
            x += perp.cos() * overrides.lateral_offset;
            y += perp.sin() * overrides.lateral_offset;
        }

        Self {
            x,
            y,
            angle,
            speed: 0.0,
            width: 28.0,
            height: 14.0,
            prev_speed: 0.0,
            turning: false,
            current_zone_name: String::new(),
            should_spawn_smoke: false,
            should_spawn_dark_smoke: false,
            name: overrides.name.unwrap_or_else(|| "Player".into()),
            color: overrides.color.unwrap_or([0.85, 0.1, 0.1]),
            is_ai: overrides.is_ai,
            physics,
            damage: DamageState::new(),
            npc: None,
        }
    }

    /// Physics step. `time` is the global elapsed time (used for flat-tire thumping).
    pub fn update(
        &mut self,
        dt: f64,
        input: &CarInput,
        track: &Track,
        time: f64,
        rng: &mut GameRng,
    ) {
        let physics = self.physics;
        let total_mass = physics.mass + physics.fuel_mass;

        let zone = track.get_surface_at(self.x, self.y);
        let (zone_grip, zone_bumpiness, zone_name) = (zone.grip, zone.bumpiness, zone.name.clone());
        self.current_zone_name = zone_name;
        let on_track = track.is_on_track(self.x, self.y);

        // Damage modifiers
        let dmg_mods = damage::get_handling_modifiers(&self.damage);

        // Tire pressure grip
        let pressure_dev = (physics.tire_pressure - physics.optimal_pressure).abs();
        let pressure_grip = (1.0 - pressure_dev * 0.4).max(0.3);

        // Effective grip (also reduced by average tire health)
        let surface_grip = if on_track { zone_grip } else { 0.3 };
        let effective_grip =
            (surface_grip * pressure_grip * physics.grip_multiplier * dmg_mods.avg_tire_health)
                .clamp(0.1, 1.0);

        // Bumpiness: suspension damage amplifies it; flat tires add periodic thump
        let base_bump = if on_track {
            zone_bumpiness * physics.bump_multiplier
        } else {
            0.0
        };
        let mut bumpiness = base_bump * dmg_mods.bump_mult;

        let flat_count = damage::flat_tire_count(&self.damage);
        if flat_count > 0 && self.speed.abs() > 15.0 {
            // Thump frequency scales with speed (like wheel hitting rim each revolution)
            let thump_freq = self.speed.abs() / 60.0;
            let thump_phase = time * thump_freq;
            let thump = (thump_phase * 2.0 * std::f64::consts::PI).sin().max(0.0);
            bumpiness += thump * flat_count as f64 * 0.35;
        }

        self.prev_speed = self.speed;

        // Throttle / brake
        let throttle = if input.up && physics.fuel_mass > 0.0 {
            1.0
        } else {
            0.0
        };
        let braking = input.down;

        // Engine damage reduces drive force
        let drive_force = throttle * physics.engine_force * effective_grip * dmg_mods.engine_mult;
        let brake_decel = if braking {
            physics.brake_force * effective_grip
        } else {
            0.0
        };

        // Drag (body damage increases drag)
        let drag_force = physics.drag_coeff * self.speed * self.speed.abs() * dmg_mods.drag_mult;

        // Rolling resistance
        let rolling_force = physics.rolling_resistance * total_mass * 9.81;

        // Off-track grass drag
        let grass_drag = if !on_track {
            self.speed.abs() * 3.0
        } else {
            0.0
        };

        // Net force
        let mut net_force = drive_force - drag_force - rolling_force - grass_drag;
        if braking {
            if self.speed > 0.0 {
                net_force -= brake_decel;
            } else if self.speed < 0.0 {
                net_force += brake_decel;
            } else {
                net_force -= brake_decel * 0.3;
            }
        }

        let accel = net_force / total_mass;
        self.speed += accel * dt;

        // Bumpiness perturbation (suspension damage amplified)
        if bumpiness > 0.01 && self.speed.abs() > 20.0 {
            let bump_mag = bumpiness * self.speed.abs() * 0.0003;
            self.speed += (rng.next_f64() - 0.5) * bump_mag * self.speed;
            self.angle += (rng.next_f64() - 0.5) * bumpiness * 0.005;
        }

        // Clamp speed (max speed reduced by damage)
        let effective_max_speed = physics.max_speed * dmg_mods.max_speed_mult;
        self.speed = self.speed.clamp(-100.0, effective_max_speed);

        // Stop drifting at low speeds
        if self.speed.abs() < 1.0 && throttle == 0.0 && !braking {
            self.speed = 0.0;
        }

        // Fuel consumption
        if throttle > 0.0 {
            self.physics.fuel_mass = (self.physics.fuel_mass - physics.fuel_rate * dt).max(0.0);
        }

        // Turning (supports both boolean left/right and continuous steer)
        let turn_factor = (self.speed.abs() / 100.0).min(1.0) * effective_grip;
        self.turning = false;
        if let Some(steer) = input.steer {
            let steer_val = steer.clamp(-1.0, 1.0);
            if steer_val.abs() > 0.05 {
                self.angle += physics.base_turn_speed * steer_val * turn_factor * dt;
                self.turning = true;
            }
        } else {
            if input.left {
                self.angle -= physics.base_turn_speed * turn_factor * dt;
                self.turning = true;
            }
            if input.right {
                self.angle += physics.base_turn_speed * turn_factor * dt;
                self.turning = true;
            }
        }

        // Tire pull: asymmetric damage pulls the car to one side
        if dmg_mods.tire_pull.abs() > 0.01 && self.speed.abs() > 10.0 {
            let pull_turn = dmg_mods.tire_pull * physics.base_turn_speed * 0.35 * turn_factor;
            self.angle += pull_turn * dt;
        }

        // Move car
        self.x += self.angle.cos() * self.speed * dt;
        self.y += self.angle.sin() * self.speed * dt;

        // Keep in bounds
        self.x = self.x.clamp(10.0, 790.0);
        self.y = self.y.clamp(10.0, 590.0);

        // Smoke flags
        let is_braking = input.down && self.speed > 50.0;
        let is_sharp_turn = self.turning && self.speed.abs() > 120.0;
        self.should_spawn_smoke = is_braking || is_sharp_turn;
        self.should_spawn_dark_smoke = self.damage.engine < 0.55 && self.speed.abs() > 25.0;
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn no_input() -> CarInput {
        CarInput::default()
    }

    fn setup() -> (Track, GameRng) {
        (Track::default_oval(), GameRng::new(42))
    }

    #[test]
    fn initializes_at_track_start_position() {
        let (track, _) = setup();
        let car = Car::new(&track, CarOverrides::default());
        assert_eq!(car.x, track.start_x);
        assert_eq!(car.y, track.start_y);
        assert_eq!(car.speed, 0.0);
        assert_eq!(car.angle, track.start_angle);
    }

    #[test]
    fn accelerates_when_throttle_pressed() {
        let (track, mut rng) = setup();
        let mut car = Car::new(&track, CarOverrides::default());
        let input = CarInput {
            up: true,
            ..no_input()
        };
        car.update(0.016, &input, &track, 0.0, &mut rng);
        assert!(car.speed > 0.0);
    }

    #[test]
    fn stays_still_with_no_input() {
        let (track, mut rng) = setup();
        let mut car = Car::new(&track, CarOverrides::default());
        car.update(0.016, &no_input(), &track, 0.0, &mut rng);
        assert_eq!(car.speed, 0.0);
    }

    #[test]
    fn turns_left_when_left_pressed_at_speed() {
        let (track, mut rng) = setup();
        let mut car = Car::new(&track, CarOverrides::default());
        car.speed = 100.0;
        let start_angle = car.angle;
        let input = CarInput {
            left: true,
            ..no_input()
        };
        car.update(0.016, &input, &track, 0.0, &mut rng);
        assert!(car.angle < start_angle);
    }

    #[test]
    fn turns_right_when_right_pressed_at_speed() {
        let (track, mut rng) = setup();
        let mut car = Car::new(&track, CarOverrides::default());
        car.speed = 100.0;
        let start_angle = car.angle;
        let input = CarInput {
            right: true,
            ..no_input()
        };
        car.update(0.016, &input, &track, 0.0, &mut rng);
        assert!(car.angle > start_angle);
    }

    #[test]
    fn decelerates_when_braking() {
        let (track, mut rng) = setup();
        let mut car = Car::new(&track, CarOverrides::default());
        car.speed = 200.0;
        let input = CarInput {
            down: true,
            ..no_input()
        };
        car.update(0.016, &input, &track, 0.0, &mut rng);
        assert!(car.speed < 200.0);
    }

    #[test]
    fn consumes_fuel_when_accelerating() {
        let (track, mut rng) = setup();
        let mut car = Car::new(&track, CarOverrides::default());
        let start_fuel = car.physics.fuel_mass;
        let input = CarInput {
            up: true,
            ..no_input()
        };
        car.update(0.016, &input, &track, 0.0, &mut rng);
        assert!(car.physics.fuel_mass < start_fuel);
    }

    #[test]
    fn does_not_accelerate_with_no_fuel() {
        let (track, mut rng) = setup();
        let mut car = Car::new(&track, CarOverrides::default());
        car.physics.fuel_mass = 0.0;
        let input = CarInput {
            up: true,
            ..no_input()
        };
        car.update(0.016, &input, &track, 0.0, &mut rng);
        assert!(car.speed <= 0.0);
    }

    #[test]
    fn speed_is_clamped_to_max_speed() {
        let (track, mut rng) = setup();
        let mut car = Car::new(&track, CarOverrides::default());
        car.speed = car.physics.max_speed + 100.0;
        car.update(0.016, &no_input(), &track, 0.0, &mut rng);
        assert!(car.speed <= car.physics.max_speed);
    }

    #[test]
    fn position_stays_within_screen_bounds() {
        let (track, mut rng) = setup();
        let mut car = Car::new(&track, CarOverrides::default());
        car.x = 5.0;
        car.y = 5.0;
        car.speed = -200.0;
        car.angle = std::f64::consts::PI;
        car.update(0.1, &no_input(), &track, 0.0, &mut rng);
        assert!(car.x >= 10.0);
        assert!(car.y >= 10.0);
    }

    #[test]
    fn applies_physics_overrides() {
        let (track, _) = setup();
        let car = Car::new(
            &track,
            CarOverrides {
                physics: PhysicsOverrides {
                    engine_force: Some(300_000.0),
                    ..Default::default()
                },
                ..Default::default()
            },
        );
        assert_eq!(car.physics.engine_force, 300_000.0);
        assert_eq!(car.physics.mass, 800.0); // other values unchanged
    }

    #[test]
    fn turns_left_with_continuous_negative_steer() {
        let (track, mut rng) = setup();
        let mut car = Car::new(&track, CarOverrides::default());
        car.speed = 100.0;
        let start_angle = car.angle;
        let input = CarInput {
            steer: Some(-0.5),
            ..no_input()
        };
        car.update(0.016, &input, &track, 0.0, &mut rng);
        assert!(car.angle < start_angle);
    }

    #[test]
    fn turns_right_with_continuous_positive_steer() {
        let (track, mut rng) = setup();
        let mut car = Car::new(&track, CarOverrides::default());
        car.speed = 100.0;
        let start_angle = car.angle;
        let input = CarInput {
            steer: Some(0.5),
            ..no_input()
        };
        car.update(0.016, &input, &track, 0.0, &mut rng);
        assert!(car.angle > start_angle);
    }

    #[test]
    fn ignores_tiny_steer_below_deadzone() {
        let (track, mut rng) = setup();
        let mut car = Car::new(&track, CarOverrides::default());
        car.speed = 100.0;
        let start_angle = car.angle;
        let input = CarInput {
            steer: Some(0.01),
            ..no_input()
        };
        car.update(0.016, &input, &track, 0.0, &mut rng);
        // Steer below deadzone (0.05) should not cause intentional turning
        // (tiny angle change may occur from bumpiness perturbation)
        assert!((car.angle - start_angle).abs() < 0.01);
    }
}
